components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,67 +85,105 @@ public AbstractHTTPFunction() {
                         String jsonString = EntityUtils.toString(response.getEntity());
                         if (contentType != null && contentType.getValue().contains(TYPE_TEXT_PLAIN)) {
-                            // For 'text/plain', put the response body into the JSON object as a single field.
                             json = new JSONObject();
                             json.put(RESPONSE, jsonString);
                         } else {
                             JSONParser parser = new JSONParser();
                             json = (JSONObject) parser.parse(jsonString);
                         }
                     }
+                    return Pair.of(outcome, json); // Success, return immediately
+                } else if (responseCode == HTTP_STATUS_UNAUTHORIZED) {
+                    LOG.warn("Received 401 response from external API call. Url: " + endpointURL);
+                    return Pair.of(Constants.OUTCOME_FAIL, null); // Unauthorized, no retry
                 } else {
-                    outcome = Constants.OUTCOME_FAIL;
+                    outcome = Constants.OUTCOME_FAIL; // Other client or server error, retry if attempts left
                 }
-
-            } catch (IllegalArgumentException e) {
-                LOG.error("Invalid Url: " + endpointURL, e);
-                outcome = Constants.OUTCOME_FAIL;
-            } catch (ConnectTimeoutException e) {
-                LOG.error("Error while waiting to connect to " + endpointURL, e);
-                outcome = Constants.OUTCOME_TIMEOUT;
-            } catch (SocketTimeoutException e) {
-                LOG.error("Error while waiting for data from " + endpointURL, e);
-                outcome = Constants.OUTCOME_TIMEOUT;
-            } catch (IOException e) {
-                LOG.error("Error while calling endpoint. ", e);
-                outcome = Constants.OUTCOME_FAIL;
-            } catch (ParseException e) {
-                LOG.error("Error while parsing response. ", e);
-                outcome = Constants.OUTCOME_FAIL;
+            } catch (Exception e) {
+                // Log the error based on its type
+                if (e instanceof IllegalArgumentException) {
+                    LOG.error("Invalid Url: " + endpointURL, e);
+                } else if (e instanceof ConnectTimeoutException) {
+                    LOG.error("Error while waiting to connect to " + endpointURL, e);
+                } else if (e instanceof SocketTimeoutException) {
+                    LOG.error("Error while waiting for data from " + endpointURL, e);
+                } else if (e instanceof IOException) {
+                    LOG.error("Error while calling endpoint. ", e);
+                } else if (e instanceof ParseException) {
+                    LOG.error("Error while parsing response. ", e);
+                }
+                outcome = Constants.OUTCOME_FAIL; // Mark as fail and check if retry is possible
```

Comment:
Just wondering whether this while loop is quite needed. Retry needs to happen only after a fail and that's when the iteration should happen.
IMO it's possible to improve this function more just taking the client invoking part to a different function and improve the overall readability

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/17a843c6983470ae35ddf181a32231b3390adbf8

Reply:
Here, what we do is that the first call itself will be invoked through the retry function.
I did that to break the code and reduce code duplications.

Reply:
I understand that's what happens but overall it's not quite readable and the method is complex

Reply:
Moreover, retries and timeouts should be recorded in diagnostics logs for troubleshooting purposes

Reply:
Fixed with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/afe4342c547dfe83345abfd23f7fee114389532b